### PR TITLE
feat: change plotly figure widget to allow image annotation

### LINKED ISSUE #249: Plotly Figure for Image Annotations
Plotly offers [image annotations features](https://dash.plotly.com/annotations). I'm building a semi automatic annotator using solara and to allow user to update the annotations I need to listen to the `relayoutData` event, an option that the solara's component doesn't offer.

Screenshot from 2023-08-21 15-45-33

I just create my custom component, but the ability to listen to `relayoutData` may be a interesting feature to add to solara. My current code:

```python
@solara.component
def FigurePlotly(
    fig,
    on_canvas_callback: Optional[Callable[[Any], None]] = None,
    dependencies=None,
):
    from plotly.graph_objs._figurewidget import FigureWidget

    fig_element = FigureWidget.element(on__js2py_relayout=on_canvas_callback)

    def update_data():
        fig_widget = solara.get_widget(fig_element)
        fig_widget.layout = fig.layout

        fig_widget.add_traces(fig.data)

    solara.use_effect(update_data, dependencies or fig)
    return fig_element
```

### PR DESCRIPTION
**Pull request by itepifanio**
Closes #249

### CODE REVIEW

File: solara/website/pages/examples/visualization/annotator.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+"""# Image annotation with Solara
+
+This example displays how to annotate images with different drawing tools in plotly figures. Use the canvas
```
Comment (maartenbreddels) [inline solara/website/pages/examples/visualization/annotator.py:3]:
```suggestion
This example displays how to annotate images with different drawing tools in plotly figures. Use the canvas
```
whitespace

File: solara/website/pages/examples/visualization/annotator.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    def on_relayout(data):
+        if data is None:
+            return
+
+        relayout_data = data['relayout_data']
+
+        if "shapes" in relayout_data:
+            text.value = str(json.dumps(relayout_data["shapes"], indent=2, cls=CustomEncoder))
+
+    with solara.Div() as main:
```
Comment (maartenbreddels) [inline solara/website/pages/examples/visualization/annotator.py:35]: I think you can remove the Div, unless you added it for a reason. and unindent the block containing it.

File: solara/website/pages/examples/visualization/annotator.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+                        "eraseshape",
+                    ]
+                }
+            )
+        )
+
+        solara.FigurePlotly(fig, on_relayout=on_relayout)
+        solara.Preformatted(text.value)
+
+    return main
```
Comment (maartenbreddels) [inline solara/website/pages/examples/visualization/annotator.py:57]:
```suggestion
```
No need anymore to return an element

**Review by maartenbreddels — CHANGES_REQUESTED** [review]
Hi Italo,

thanks for your PR, really nice work. Mostly small comments, nothing major. I like the refactor you've done on the on_points_callback 👍 

Regards,

Maarten

File: solara/components/misc.py [hunk 4]
```diff
@@ -272,39 +272,41 @@ def FigurePlotly(
...
         fig_widget.layout = fig.layout
-
-        length = len(fig_widget.data)
         fig_widget.add_traces(fig.data)
-        data = list(fig_widget.data)
-        fig_widget.data = data[length:]
+        if hasattr(fig_widget, 'data'):
+            length = len(fig_widget.data)
+            data = list(fig_widget.data)
+            fig_widget.data = data[length:]
```
[the remark this answers is not in the document]
Comment (itepifanio) [reply solara/components/misc.py]: Reverted. Some changes I made in my codebase was added here by mistake, I was testing some things on the configuration. Sorry about that.

File: solara/website/pages/examples/visualization/annotator.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+Check [plotly docs](https://dash.plotly.com/annotations) for more information about image annotation.
+"""
+import json
+import plotly.graph_objects as go
+
+import solara
+
+title = "Plotly Image Annotator"
+text = solara.reactive('Draw on canvas')
```
[the remark this answers is not in the document]
Comment (itepifanio) [reply solara/website/pages/examples/visualization/annotator.py]: Done

File: solara/website/pages/__init__.py [hunk 6]
```diff
@@ -95,7 +96,7 @@ def Layout(children=[]):
                 sub_title="Solara helps you build powerful & scalable Jupyter and web apps <b>faster</b> and <b>easier</b>.",
                 button_text="Quickstart",
             )
-
+"""
```
Comment (maartenbreddels) [inline solara/website/pages/__init__.py:99]: why was this needed? Did you have trouble running this part? Is there something can do about it?
In any case, for this PR I think you want to revert this.

File: solara/website/pages/examples/visualization/annotator.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+title = "Plotly Image Annotator"
+shapes = solara.reactive(None)
+
+class CustomEncoder(json.JSONEncoder):
+    def default(self, o):
+        # Convert object instances to their string representation
+        if isinstance(o, object):
+            return str(o)
+        return super().default(o)
```
Comment (maartenbreddels) [inline solara/website/pages/examples/visualization/annotator.py:22]: Why is this needed? Does plotly give back some objects? If so, it would be good to explain this in this piece of code.

**Review by maartenbreddels — CHANGES_REQUESTED** [review]
Thank you for the changes.
I see two files which are removed, please add them back, and please take a look at the comments I gave.
Hope we can get this in soon!

File: solara/website/pages/__init__.py (hunk 6 above)
Comment (itepifanio) [reply solara/website/pages/__init__.py:99]: @maartenbreddels yes, I had issues running the documentation* and ended up removing this files to run the documentation. This wasn't suppose to be deleted, it should be revered now.

*[I followed the setup docs](https://solara.dev/docs/development) but had problems running the tests and docs. I'll try to setup again later and open some issues if needed.

File: solara/website/pages/examples/visualization/annotator.py (hunk 7 above)
Comment (itepifanio) [reply solara/website/pages/examples/visualization/annotator.py]: > Does plotly give back some objects?

Yes, I added more explanation about it.

**Comment by maartenbreddels** [comment]
Did some small changes, such as making the annotation mode the default.

Thank you a lot! ❤️
